### PR TITLE
77

### PR DESCRIPTION
**Pull request by danse**

### CODE REVIEW

**Comment by danse** [comment]
Hey @skoni, while i am still testing this and i might add some bug fixes, i think that it is ready to be merged, so feel free to go through the code and point the things you do not like. These are the problems i still experience with this:
- [x] changes feed timeout. this might be due to the gateways i have between my machine and the database. anyway changes feeds are wrapped into a `Bacon.repeat` which makes them robust to this
- [x] the send function `sendToMobile` in `lib2.js` takes more time than expected. this is a very simple function so i am not sure about what is going on here, i will test it now

**Comment by danse** [comment]
This is a simple one-liner which can be used to test the send function:

```
require('./lib2.js').withOptions({ database: 'test', gateway: 'https://developers:XXX@dev-sl-ebola-cases.eocng.org/_couchdb/sense/_messages', debug:true }).sendToMobile({ content:'test', to: '1234' });
```

**Comment by danse** [comment]
great! the problem with that function seems fixed now that i installed a recent version of `request`!

**Comment by danse** [comment]
rebased! i think that this is ready to be merged

File: README.md [hunk 1]
```diff
@@ -2,6 +2,49 @@
...
+existing document.
+
+#### What about the 2?
+
+I was working on this project after one year of inactivity, and for a
+different use case. New functionalities i needed were all easier to
+write from scratch, but i can't exclude that we might want to use the
+preexisting logic, so i created some new files marked with `2`, namely
+a new library, `lib2` and a new executable,
+`sense-dispatch-2`. Eventually this duplication could be eliminated.
```
Comment (tillre) [inline README.md]: It would be nice to have an example of the cli invocation.

Comment (danse) [inline README.md]: sure i need to extend the doc and mention also the conf document, i will add something now

Comment (danse) [inline README.md]: i was waiting to deploy because i expected that deployment would require me to change some detail

**Comment by tillre** [comment]
I find the project layout confusing. Can we get rid of the old files? I dont think they will ever be used again and `rename sense-dispatch-2.js` to `index.js` or something obvious. It takes some looking around to understand whats going on and if you also think the old stuff is obsolete, then lets restructure this. The former version is in the history.

**Comment by danse** [comment]
@skoni i agree, what do you think that i should do with `lib.js`?

**Comment by tillre** [comment]
If its not used, delete it.

**Comment by danse** [comment]
alright

**Comment by tillre** [comment]
i have problems to get it running, trying like this:

```
node sense-dispatch-2.js -d http://localhost:5984/sense -g http://localhost:7362 -b
error fetching the configuration document
{"promise error":"{\"0\":{\"status\":404,\"name\":\"not_found\",\"message\":\"missing\",\"error\":true,\"reason\":\"missing\"}}"}
```

couchdb log:

```
[info] [<0.102.0>] 127.0.0.1 - - GET /sense/ 200
[info] [<0.114.0>] 127.0.0.1 - - GET /sense/sense-dispatch-configuration 404
```

The url to the relerivet gateway is wrong, but i think its trying to find the config-doc which is not there. I will try with a proper url for the gateway.

**Comment by danse** [comment]
Yes you need to add the conf document. I added some doc right now. An example of conf doc is https://dev-sl-ebola-cases.eocng.org/_couchdb/_utils/document.html?sense/sense-dispatch-configuration

**Comment by danse** [comment]
hey i removed the legacy files as you said :sunglasses:

File: README.md [hunk 2]
```diff
@@ -2,6 +2,45 @@
...
+### Command line use
+
+An example use is:
+
+    sense-dispatch.js -b -g https://user:pass@dev-sl-ebola-cases.eocng.org/_couchdb/sense/_messages -d https://user:pass@dev-sl-ebola-cases.eocng.org/_couchdb/sense
+
+Note that `-b` activates extra debugging info and can be omitted.
+
+An important part of the configuration will be fetched from the
+database, in the document with id `sense-dispatch-configuration`. This
```
Comment (tillre) [inline README.md]: Saying explicitly that the user has to upload the document first would make it more clear.

**Comment by tillre** [comment]
Atm the service will abort if it does not find the configuration doc. It would then be restarted a couple times before supervisor gives up on it. So if you add the config doc after the server has been deployed, the service would have to be manually restarted. You could not fail if the document is not there and instead check for it every X seconds until it finds it. What do you think, just thinking about it.

**Comment by danse** [comment]
> You could not fail if the document is not there and instead check for it
> every X seconds until it finds it. What do you think, just thinking about it

This is one of the few parts which is not wrapped in Bacon, but in a
promise, so it is not easy to retry. Now from the top of my head i would
say that it is not worth the complexity in the code. I am retrying things
that are necessary, like gateway timeouts, or changes in configuration.
Unfortunately this service is not robust against misconfigurations yet.

File: configuration-examples/1.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  recipients: [],
```
Comment (tillre) [inline configuration-examples/1.json]: this is missing qoutes to be valid json

**Comment by tillre** [comment]
While we start developing more services it would be good to have a common layout and readme. Currently this https://github.com/eHealthAfrica/migration-followup-separate-followups and dispatch differ quite some. Its not hard to read how something is done, but nontheless it makes the life easier for people who just integrate something. cc @jo

**Comment by tillre** [comment]
nice :bowtie:

**Comment by danse** [comment]
:tada:

**Comment by jo** [comment]
@skoni I agree that a commom README layout for services would be very useful.

Normally I structure my READMEs like that:
1. Name
2. Description
3. Code snippet or usage example
4. Installation
5. Detailed usage
6. Testing
7. License

We might want to dig into this topic a little further. [awesome-readme](https://github.com/matiassingers/awesome-readme) lists some examples of good READMEs as well as some articles about writing them.

**Comment by tillre** [comment]
That list looks good to me. We dont have to overdo it. I would say lets stick with the points you worte down. Is there a place where we keep such information?

**Comment by jo** [comment]
Propably http://docs.ehealthafrica.org/

**Pull request merged**